libcudacxx/include/cuda/std/__iterator/access.h

```diff
@@ -21,47 +21,101 @@
 #  pragma system_header
 #endif // no system header
 
+#include <cuda/std/__type_traits/enable_if.h>
+#include <cuda/std/__type_traits/is_same.h>
+#include <cuda/std/__utility/declval.h>
 #include <cuda/std/cstddef>
+#include <cuda/std/initializer_list>
 
 _LIBCUDACXX_BEGIN_NAMESPACE_STD
 
+// We need to detect whether there is already a free function begin / end that would end up being ambiguous.
+// This can happen when a type pulls in both namespace std and namespace cuda::std via ADL.
+// In that case we are always safe to just not do anything because that type must be host only.
+namespace __detect_ambiguous_access
+{
+struct __not_ambiguous
+{};
+template <class _Tp, size_t _Np>
+_LIBCUDACXX_INLINE_VISIBILITY auto __begin(_Tp (&__array)[_Np]) -> decltype(begin(__array));
+template <class _Cont>
+_LIBCUDACXX_INLINE_VISIBILITY auto __begin(_Cont& __c) -> decltype(begin(__c));
+template <class _Cont>
+_LIBCUDACXX_INLINE_VISIBILITY auto __begin(const _Cont& __c) -> decltype(begin(__c));
+_LIBCUDACXX_INLINE_VISIBILITY auto __begin(...) -> __not_ambiguous;
+template <class _Cont>
+struct __begin_not_ambiguous
+    : is_same<decltype(__detect_ambiguous_access::__begin(_CUDA_VSTD::declval<_Cont>())), __not_ambiguous>
```

Comment:
Suggestion: This meta function does not check whether `__begin(Cont)` is ambigous. It checks whether there is a reachable overload. So I would name it differently, like: `has___begin_overload_for`.

Then the constraint later would read:
```c++
template <class _Tp, size_t _Np>
__enable_if_t<!__detect_ambiguous_access::has___begin_overload_for<_Tp (&)[_Np]>::value, _Tp*>
```
Which is a tiny improvement in readability.

Reply:
:+1: I went with `__has_no_adl_meow`